### ISSUE TITLE
Use native PowerShell progress bar instead of ugly shell simulated progress bar

### ISSUE THREAD

**Issue by soroshsabz** [microsoft/winget-cli#1429]
# Description of the new feature/enhancement

Replace PowerShell progress bar instead of ugly shell based progress simulated when we try download to install package in PowerShell with `winget`

I think it is good to replace below

![image](https://user-images.githubusercontent.com/17947618/131961160-fca718a9-ebd6-4f7b-b6bc-1476fe3f47d7.png)

with

![image](https://user-images.githubusercontent.com/17947618/131961283-832cd00d-0953-4f74-8d33-eb813f7c183b.png)

# Proposed technical implementation details (optional)

I suggest using [write-progress](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/write-progress?view=powershell-7.1) command when using `winget` in PowerShell to make better user experience in PowerShell eco-system and feel in-house and native command in PowerShell

This [blog](https://jordantheitguy.com/2019/06/24/how-to-powershell-progress-bars/) is very helpful for writing progress bar in powershell, and this progress can easily show percentage, bytes remaining and **speed**( this is very needed for downloading from internet with limited speed)

**Comment by soroshsabz** [comment]
related to #221 

related to #674 

related to #1088

**Comment by jedieaston** [comment]
This needs to be user configurable. The PowerShell progress bar (at least in 5.1) is buggy as heck whenever I try to use it, so most of the time I suppress it. The winget progress bar works fine, on the other hand.

**Comment by denelon** [comment]
This would be configurable just like "rainbow", "retro", etc.

**Comment by jantari** [comment]
For what it's worth the PowerShell-native progress bar has been changed to look pretty similar to the current one winget presents in PowerShell 7.2

I personally don't think this is a pressing issue as progress bars are a UI feature for interactive users (humans) and that's not really how I'm going to use winget - but if it's not too much work, sure.

**Comment by soroshsabz** [comment]
@jantari I think it is important to Microsoft ecosystem is being consistent, so I think all Microsoft CLI app must be used native PowerShell progress bar if the app need progress bar, to achieve same UX in Microsoft ecosystem.

thanks

**Comment by VivaLosDoyers** [comment]
What about for those us that don't use PowerShell though? Keep this is a fallback?

**Comment by rdeforest** [comment]
For those of us who think the default PowerShell progress bar is ugly, we can at least improve the colors by putting something like this in ~/Documents/WindowsPowerShell/profile.ps1

```
$Host.PrivateData.ProgressForegroundColor='White'
$Host.PrivateData.ProgressBackgroundColor='Blue'
```

I still strongly favor the winget progress bar.

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
> http://github.com/microsoft/winget-cli/issues/1429#issuecomment-931639466;
> http://github.com/microsoft/winget-cli/issues/1429#issuecomment-912611004.

If that is problematic for you, report that to the maintainers of PowerShell, because if those problems are remediated, this software shall gain consistent indicators of progress, and PowerShell's shall have been improved.

**Comment by jedieaston** [comment]
It's a problem in PowerShell 5.1, as far as I can tell it's fine (or at least, better) in PowerShell 7. I don't think PowerShell 5.1 is being supported for anything other than emergency bug fixes, or else I would (although I'm sure they're aware of it).

**Comment by rdeforest** [comment]
> > [http://github.com/microsoft/winget-cli/issues/1429#issuecomment-931639466](https://github.com/microsoft/winget-cli/issues/1429#issuecomment-931639466), and [http://github.com/microsoft/winget-cli/issues/1429#issuecomment-912611004](https://github.com/microsoft/winget-cli/issues/1429#issuecomment-912611004).
> 
> If that is problematic for you, report that to the maintainers of PowerShell, because if those problems are remediated, this software shall gain consistent indicators of progress, and PowerShell's shall have been improved.

Yes, improving PowerShell is a great idea, but this issue #1429 is about the winget-cli project. Improving the PS progress bar is out of scope.

When there is no feature gap between the progress bars in `winget` and PowerShell then this issue will have some legs. Until then it should be maximally de-prioritized in favor of work which will make `winget` better.

I don't speak for the winget-cli team, I'm just a person on the Internet with opinions.

**Comment by Masamune3210** [comment]
Oh, definitely fix the actual issues first

**Comment by Karl-WE** [comment]
Ihmo it looks prettier in Microsoft Terminal and PowerShell 7.x

**Comment by Spacellary** [comment]
Just feedback from a regular user here, but I'd much prefer a mix of both, but more so a PowerShell 7.x styled one but in Gray or B/W.

**Comment by rokejulianlockhart** [comment]
@Spacellary, I believe that Write-Progress accepts colouration parameters.

**Comment by davidldennison** [comment]
Add this to your settings.json config file for Winget to get a nice rainbow version

```
"visual": {
        "progressBar": "rainbow",
```

![image](https://github.com/microsoft/winget-cli/assets/67825086/fbfae9c0-6129-46d2-8576-0c5c1c1e5c69)

**Comment by Karl-WE** [comment]
@soroshsabz  is this still something you pursue or is the one from @davidldennison plus the Windows Terminal View in PowerShell 7.x enough? 
Thank you for considering housekeeping and close if no longer requested.

**Comment by soroshsabz** [comment]
As @RokeJulianLockhart said, I think winget-cli can implement progress bar with native PowerShell progress bar feature and I think it is very odd to winget-cli does not support native PowerShell Progress Bar

**Comment by Karl-WE** [comment]
I see your point clearer now. Can just believe they do not want to use PowerShell progress bar to reduce dependencies. 
imagine that something in progress bar function changes and then it would break. as PS is not maintained by winget team, you might understand the complexity. in addition, PS 5.1, 7 or later might have different syntax. this is usually the case for other commands aswell. Not speaking for winget team, just bringing up my thoughts why they might have chosen this way.
When winget has been released, it had no PowerShell support.
+1 for consistency :)

**Comment by soroshsabz** [comment]
@denelon Any update for this?

thanks :)

**Comment by denelon** [comment]
**Issue summary**
- Request: have `winget` render the native PowerShell progress bar (`Write-Progress`) instead of its own shell-drawn progress bar when run in PowerShell.
- Goal: a progress experience that feels native and consistent within the PowerShell ecosystem.

**Disposition:** Not planned.

**Why:** `winget.exe` is a native command-line executable, not a PowerShell cmdlet. `Write-Progress` is a PowerShell host construct that only exists inside a PowerShell runspace, so the CLI cannot call it — when `winget` runs under PowerShell it is simply a child process writing to the console, exactly as it does under Command Prompt, Windows Terminal, SSH, or CI. The CLI is intentionally terminal-agnostic and must render the same way across all of these hosts, so it can't depend on any single shell's progress API.

The native PowerShell progress experience this issue asks for already exists in the **`Microsoft.WinGet.Client` PowerShell module**: cmdlets such as `Install-WinGetPackage` and `Update-WinGetPackage` report progress through PowerShell's native `Write-Progress` pipeline. If you want progress that integrates natively with PowerShell, use the module.

For the CLI itself, the progress bar is already configurable and adapts to the terminal. You can change its style in `settings.json`:

```json
"visual": {
    "progressBar": "accent"   // also: "rainbow", "retro", or "disabled"
}
```

It also automatically falls back to a plain, non-VT rendering on terminals that don't support virtual terminal sequences.

Given that the native-PowerShell experience is provided by the module and the CLI bar is already configurable and host-aware, we're closing this as not planned.

*Drafted with the help of GitHub Copilot.*
